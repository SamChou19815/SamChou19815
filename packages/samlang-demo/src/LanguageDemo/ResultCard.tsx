import React, { ReactNode, ReactElement } from 'react';

import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import CodeBlock from 'lib-react/PrismCodeBlock';

import DemoStyles from './LanguageDemo.module.css';
import ResultStyles from './ResultCard.module.css';
import { Response } from './interpret';

type Props = {
  readonly response: Response | 'waiting' | 'server-error' | null;
};

const ErrorDetail = ({ children }: { readonly children: string }): ReactElement => (
  <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
    <code>
      {children.split('\n').map((line, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <p key={index}>{line}</p>
      ))}
    </code>
  </div>
);

/**
 * The component of the language demo.
 */
export default function ResultCard({ response }: Props): ReactElement {
  let children: ReactNode;
  if (response === null) {
    children = 'Submit a program to see the interpretation result.';
  } else if (response === 'waiting') {
    children = 'Waiting for server response...';
  } else if (response === 'server-error') {
    children = 'Server Error. Please try again.';
  } else {
    switch (response.type) {
      case 'GOOD_PROGRAM': {
        const { result, prettyPrintedProgram } = response.detail;
        const resultNodeStyle = {
          borderLeftColor: result.startsWith('Value') ? 'green' : 'orange',
        };
        children = (
          <div>
            <div className={ResultStyles.ColoredResult} style={resultNodeStyle}>
              <h3>Program Running Result</h3>
              <pre>{result}</pre>
            </div>
            <div>
              <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'green' }}>
                <h3>Pretty Printed Program:</h3>
                <CodeBlock language="samlang" className={ResultStyles.CodeBlock}>
                  {prettyPrintedProgram}
                </CodeBlock>
              </div>
            </div>
          </div>
        );
        break;
      }
      case 'BAD_SYNTAX':
        children = <ErrorDetail>{response.detail}</ErrorDetail>;
        break;
      case 'BAD_TYPE':
        children = <ErrorDetail>{response.detail}</ErrorDetail>;
        break;
      default:
        throw new Error('Bad Response');
    }
  }
  return (
    <Card className={DemoStyles.ParallelCard}>
      <CardHeader title="Interpretation Result" />
      <CardContent>{children}</CardContent>
    </Card>
  );
}
