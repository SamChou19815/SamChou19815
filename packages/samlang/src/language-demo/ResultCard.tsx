/**
 * Copyright (c) 2019-present, Developer Sam.
 *
 * This source code is licensed under the AGPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import CodeBlock from '@theme/CodeBlock';
import clsx from 'clsx';
import React, { ReactNode, ReactElement } from 'react';

import DemoStyles from './LanguageDemo.module.css';
import ResultStyles from './ResultCard.module.css';
import type { Response } from './demo';

const ErrorDetail = ({ children }: { readonly children: readonly string[] }): ReactElement => (
  <div className={`${ResultStyles.ColoredResult} ${ResultStyles.BadResult}`}>
    <h3>Compile Time Errors:</h3>
    <CodeBlock>{children.join('\n')}</CodeBlock>
  </div>
);

const JSBlock = ({ children }: { readonly children: string }): ReactElement => (
  <div className={`${ResultStyles.ColoredResult} ${ResultStyles.NeutralResult}`}>
    <h3>Compiled JS:</h3>
    <CodeBlock className="javascript">{children.trim()}</CodeBlock>
  </div>
);

const LLVMBlock = ({ children }: { readonly children: string }): ReactElement => (
  <div className={`${ResultStyles.ColoredResult} ${ResultStyles.NeutralResult}`}>
    <h3>Optimized LLVM Code:</h3>
    <CodeBlock className="llvm">{children.trim()}</CodeBlock>
  </div>
);

type Props = { readonly response: Response | string };

/** The component of the language demo result. */
export default function ResultCard({ response }: Props): ReactElement {
  let children: ReactNode;
  if (typeof response === 'string') {
    children = (
      <div>
        <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
          <h3>Interpreter Error</h3>
          <code>{response}</code>
        </div>
      </div>
    );
  } else {
    const { interpreterPrinted, prettyPrintedProgram, jsString, llvmString, errors } = response;
    children = (
      <div>
        {interpreterPrinted && (
          <div className={`${ResultStyles.ColoredResult} ${ResultStyles.GoodResult}`}>
            <h3>Program Standard Out:</h3>
            <CodeBlock>{interpreterPrinted}</CodeBlock>
          </div>
        )}
        {prettyPrintedProgram && (
          <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'green' }}>
            <h3>Pretty Printed Program:</h3>
            <CodeBlock className="samlang">{prettyPrintedProgram.trim()}</CodeBlock>
          </div>
        )}
        {jsString && <JSBlock>{jsString}</JSBlock>}
        {llvmString && <LLVMBlock>{llvmString}</LLVMBlock>}
        {errors.length > 0 && <ErrorDetail>{errors}</ErrorDetail>}
      </div>
    );
  }
  return (
    <div className={clsx('card', DemoStyles.ParallelCard)}>
      <div className="card__body">{children}</div>
    </div>
  );
}
