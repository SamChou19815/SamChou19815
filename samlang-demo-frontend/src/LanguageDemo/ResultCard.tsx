import React, { ReactNode, ReactElement } from 'react';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import CodeBlock, { initialize } from 'sam-highlighter';
import { Response } from './interpret';
import DemoStyles from './LanguageDemo.module.css';
import ResultStyles from './ResultCard.module.css';

type Props = {
  readonly response: Response | 'waiting' | 'server-error' | null;
};

initialize();

/**
 * The component of the language demo.
 */
export default function ResultCard({ response }: Props): ReactElement {
  let children: ReactNode;
  if (response === null) {
    children = 'Submit a program to see the interpretation result.';
  } else if (response === 'waiting') {
    children = 'Waiting for server response...';
  } else if (response === 'server-error') {
    children = 'Server Error. Please try again.';
  } else {
    switch (response.type) {
      case 'GOOD_PROGRAM': {
        const { result, prettyPrintedProgram } = response.detail;
        const resultNodeStyle = {
          borderLeftColor: result.startsWith('Value') ? 'green' : 'orange',
        };
        children = (
          <div>
            <div className={ResultStyles.ColoredResult} style={resultNodeStyle}>
              <h3>Program Running Result</h3>
              <code>{result}</code>
            </div>
            <div>
              <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'green' }}>
                <h3>Pretty Printed Program:</h3>
                <CodeBlock language="samlang">{prettyPrintedProgram}</CodeBlock>
              </div>
            </div>
          </div>
        );
        break;
      }
      case 'BAD_SYNTAX':
        children = (
          <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
            <code>{response.detail}</code>
          </div>
        );
        break;
      case 'BAD_TYPE':
        children = (
          <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
            <code>{response.detail}</code>
          </div>
        );
        break;
      default:
        throw new Error('Bad Response');
    }
  }
  return (
    <Card className={DemoStyles.ParallelCard}>
      <CardHeader title="Interpretation Result" />
      <CardContent>{children}</CardContent>
    </Card>
  );
}
