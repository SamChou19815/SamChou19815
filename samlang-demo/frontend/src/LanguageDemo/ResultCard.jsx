// @flow strict

import type { Node } from 'react';
import React from 'react';
// $FlowFixMe
import Card from '@material-ui/core/Card';
// $FlowFixMe
import CardContent from '@material-ui/core/CardContent';
// $FlowFixMe
import CardHeader from '@material-ui/core/CardHeader';
import type { Response } from './interpret';
import DemoStyles from './LanguageDemo.module.css';
import ResultStyles from './ResultCard.module.css';

type Props = {|
  +response: Response | 'waiting' | 'server-error' | null;
|};

/**
 * The component of the language demo.
 */
export default function ResultCard({ response }: Props): Node {
  let children: Node;
  if (response === null) {
    children = 'Submit a program to see the interpretation result.';
  } else if (response === 'waiting') {
    children = 'Waiting for server response...';
  } else if (response === 'server-error') {
    children = 'Server Error. Please try again.';
  } else {
    switch (response.type) {
      case 'GOOD_PROGRAM': {
        const { result, prettyPrintedProgram } = response.detail;
        const resultNodeStyle = {
          borderLeftColor: result.startsWith('Value') ? 'green' : 'orange',
        };
        children = (
          <div>
            <div className={ResultStyles.ColoredResult} style={resultNodeStyle}>
              <h3>Program Running Result</h3>
              <code>{result}</code>
            </div>
            <div>
              <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'green' }}>
                <h3>Pretty Printed Program:</h3>
                <code><pre>{prettyPrintedProgram}</pre></code>
              </div>
            </div>
          </div>
        );
        break;
      }
      case 'BAD_SYNTAX':
        children = (
          <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
            <code>{response.detail}</code>
          </div>
        );
        break;
      case 'BAD_TYPE':
        children = (
          <div className={ResultStyles.ColoredResult} style={{ borderLeftColor: 'red' }}>
            <code>{response.detail}</code>
          </div>
        );
        break;
      default:
        throw new Error('Bad Response');
    }
  }
  return (
    <Card className={DemoStyles.ParallelCard}>
      <CardHeader title="Interpretation Result" />
      <CardContent>{children}</CardContent>
    </Card>
  );
}
