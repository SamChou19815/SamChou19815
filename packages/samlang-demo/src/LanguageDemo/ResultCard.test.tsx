import React from 'react';
import renderer from 'react-test-renderer';
import ResultCard from './ResultCard';

it('ResultCard(GOOD_PROGRAM) matches snapshot.', () => {
  const response = {
    type: 'GOOD_PROGRAM',
    detail: { result: 'haha', prettyPrintedProgram: 'class HelloWorld {}' }
  };
  const tree = renderer.create(<ResultCard response={response} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ResultCard(BAD_SYNTAX) matches snapshot.', () => {
  const tree = renderer
    .create(<ResultCard response={{ type: 'BAD_SYNTAX', detail: 'foo' }} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('ResultCard(BAD_TYPE) matches snapshot.', () => {
  const tree = renderer
    .create(<ResultCard response={{ type: 'BAD_TYPE', detail: 'bar' }} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('ResultCard(waiting) matches snapshot.', () => {
  const tree = renderer.create(<ResultCard response="waiting" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ResultCard(server-error) matches snapshot.', () => {
  const tree = renderer.create(<ResultCard response="server-error" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ResultCard(null) matches snapshot.', () => {
  const tree = renderer.create(<ResultCard response={null} />).toJSON();
  expect(tree).toMatchSnapshot();
});
