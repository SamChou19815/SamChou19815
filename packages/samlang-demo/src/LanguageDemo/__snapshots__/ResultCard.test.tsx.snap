// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultCard(BAD_SYNTAX) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <div
      className="ColoredResult"
      style={
        Object {
          "borderLeftColor": "red",
        }
      }
    >
      <code>
        <p>
          foo
        </p>
      </code>
    </div>
  </div>
</div>
`;

exports[`ResultCard(BAD_TYPE) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <div
      className="ColoredResult"
      style={
        Object {
          "borderLeftColor": "red",
        }
      }
    >
      <code>
        <p>
          bar
        </p>
      </code>
    </div>
  </div>
</div>
`;

exports[`ResultCard(GOOD_PROGRAM) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    <div>
      <div
        className="ColoredResult"
        style={
          Object {
            "borderLeftColor": "orange",
          }
        }
      >
        <h3>
          Program Running Result
        </h3>
        <code>
          haha
        </code>
      </div>
      <div>
        <div
          className="ColoredResult"
          style={
            Object {
              "borderLeftColor": "green",
            }
          }
        >
          <h3>
            Pretty Printed Program:
          </h3>
          <pre>
            <code
              className="samlang"
            >
              class HelloWorld {}
            </code>
          </pre>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ResultCard(null) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    Submit a program to see the interpretation result.
  </div>
</div>
`;

exports[`ResultCard(server-error) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    Server Error. Please try again.
  </div>
</div>
`;

exports[`ResultCard(waiting) matches snapshot. 1`] = `
<div
  className="MuiPaper-root MuiCard-root ParallelCard MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
      >
        Interpretation Result
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root"
  >
    Waiting for server response...
  </div>
</div>
`;
